### ISSUE TITLE
lerna bootstrap modifies package-lock.json files

### ISSUE THREAD

**Issue by yoavain-sundaysky** [lerna/lerna#2447]
Our project have several packages:
A
B
C depends on A & B
D depends on A, B & C

all packages have package-lock.json file.

When running lerna --bootstrap

## Expected Behavior
symbolic links should be created.
package-lock.json should not be modified (only when running lerna publish, package-lock.json should be modified)

## Current Behavior
symbolic links are created.
package-lock.json files are modified, removing references of all local packages.

Workaround:
running
`lerna exec --concurrency 1 git checkout -- package-lock.json`
right after lerna bootstrap to correct package-lock.json files

 Environment info:

  System:
    OS: Windows 10 10.0.17134
    CPU: (8) x64 Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
  Binaries:
    Node: 12.15.0 - C:\Program Files\nodejs\node.EXE
    npm: 6.13.7 - C:\Program Files\nodejs\npm.CMD
  Utilities:
    Git: 2.25.0.
  npmPackages:
    lerna: 3.20.2 => 3.20.2

**Comment by nlaffey** [comment]
I'm experiencing this issue as well. I am however suing the hoist flag as well: `lerna bootstrap --hoist`. Subscribing as I'm curious if there's a way I can prevent updating my entire package-lock.json file every time I add a dependency in any of my packages.

**Comment by zgriesinger** [comment]
We have encountered a similar "issue".

It seems the `lerna bootstrap` command is functioning as intended. In our case we have this scenario:
- Using `npm`
- Using `lerna bootstrap --hoist`
- Managing private packages that are _not_ pushed to remote
- 1 package-lock file at root (Because dependencies are hoisted)

When cloning the repo, we were having users run `npm install`. The issue with this, is that it will completely rewrite the root `package-lock.json` file to only have the dependencies specified at the root. This removes the entries written to the lock file with `lerna bootstrap --hoist`. Then, when a user runs `lerna bootstrap --hoist` they get every single dependency with a `^` in the version updated.

Once we started omitting the `npm install` step and requiring a pure clone to use `npx lerna bootstrap --hoist` to pull down dependencies, our `package-lock.json` file started to look like we would expect, with only new additions being new added dependencies.

However, this is a little bit unorthodox. Having a repository using `npm` but requiring users to _not_ run `npm install` is weird for new users. We have added this to prevent someone from inadvertently running `npm install`: 
```
 "preinstall": "echo 'Please use `npx lerna bootstrap` to install dependencies' && exit 1",
```

Using `yarn` as a package manager somewhat gets around this, because the bootstrap step and the initial install step uses the same (`yarn`) binary. We haven't switched to yarn because it doesn't provide an easy way to run a scoped bootstrap when dependencies are private and not published to any registry.

I'd be interested to know if there's any way to maintain the ability to to a pure `npm install` without rewriting the root `package-lock` and thus causing `lerna bootstrap --hoist` to upgrade all `^` dependencies.

**Comment by yoavain** [comment]
> We have encountered a similar "issue".
> 
> It seems the `lerna bootstrap` command is functioning as intended. In our case we have this scenario:
> 
> * Using `npm`
> * Using `lerna bootstrap --hoist`
> * Managing private packages that are _not_ pushed to remote
> * 1 package-lock file at root (Because dependencies are hoisted)
> 
> When cloning the repo, we were having users run `npm install`. The issue with this, is that it will completely rewrite the root `package-lock.json` file to only have the dependencies specified at the root. This removes the entries written to the lock file with `lerna bootstrap --hoist`. Then, when a user runs `lerna bootstrap --hoist` they get every single dependency with a `^` in the version updated.
> 
> Once we started omitting the `npm install` step and requiring a pure clone to use `npx lerna bootstrap --hoist` to pull down dependencies, our `package-lock.json` file started to look like we would expect, with only new additions being new added dependencies.
> 
> However, this is a little bit unorthodox. Having a repository using `npm` but requiring users to _not_ run `npm install` is weird for new users. We have added this to prevent someone from inadvertently running `npm install`:
> 
> ```
>  "preinstall": "echo 'Please use `npx lerna bootstrap` to install dependencies' && exit 1",
> ```
> 
> Using `yarn` as a package manager somewhat gets around this, because the bootstrap step and the initial install step uses the same (`yarn`) binary. We haven't switched to yarn because it doesn't provide an easy way to run a scoped bootstrap when dependencies are private and not published to any registry.
> 
> I'd be interested to know if there's any way to maintain the ability to to a pure `npm install` without rewriting the root `package-lock` and thus causing `lerna bootstrap --hoist` to upgrade all `^` dependencies.

I'm not sure it's the same scenario. When you add the "--hoist" flag, you should expect changes to your lock files, as dependencies are moved to the project root.
I;m running without the "--hoist". I expect the "lerna bootstrap" command to install all dependencies in all packages, and replace the "managed" packages in with a symbolic link where ever it is possible i.e. pure file system change.

Regarding the ^ you see that is being added, perhaps you can add "save-exact=true" to your .npmrc file (in every level you have a package.json file.

**Comment by zgriesinger** [comment]
Yeah you are correct, my issue is more similar to @nlaffey's comment above

We are using the save-exact and exact: true flags.

**Comment by raijinsetsu** [comment]
`lerna bootstrap` is just calling `npm install` under the hood after removing local packages from the package.json. So, expect your package-lock.json file to change per npm's algorithm.

Use `lerna bootstrap --ci` to prevent the package-lock.json changes.

**Comment by evertonfraga** [comment]
@zgriesinger Hi! that makes a lot of sense. To help onboarding contributors and developers, I am running `lerna bootstrap` as a `postinstall` hook (see our [package.json](https://github.com/ethereumjs/ethereumjs-vm/blob/master/package.json#L14)).

To also help devs, [I described each npm script](https://github.com/ethereumjs/ethereumjs-vm#project-scripts--run-from-repository-root), making a big warning about using npm install on the `./packages/xxx`, as it can mess with the environment and lead to frustration.

**Comment by yoavain-sundaysky** [comment]
> `lerna bootstrap` is just calling `npm install` under the hood after removing local packages from the package.json. So, expect your package-lock.json file to change per npm's algorithm.
> 
> Use `lerna bootstrap --ci` to prevent the package-lock.json changes.

The `--ci` flag works, but it means every time someone runs `npm install`, it does a clean install. This is problematic in a large scale project.

**Comment by fyyyyy** [comment]
We have the issue that some of the modules inside a package may run `npm i` outside of our control. E.g. a user runs tests, and stencil test is installing dependencies and completely screws up the mono repo.

**Comment by thinkkevin** [comment]
I have some problem with this recently as well. 

Think about a team is working a lerna repo, someone created a new module using `react-intl` and merged into master branch with a stable package-lock.json, which was generated with `npm install` and `lerna bootstrap --hoist`.
 
After a few days, some open source libraries (e.g. `@formatjs/intl-numberformat`) updated, another team member checkout the master branch and run `npm install` with `lerna bootstrap --hoist`. He saw a changed package-lock.json. He got confused whether he should commit the change, and when he worked on his own module to introduce other dependencies but the latest change on `@formatjs/intl-numberformat` would also be brought into package-lock.json, which is un-expected.

Another challenge related to that, after I use `lerna add` to add dependency to a package, I have to run `lerna boostrap --hoist` otherwise it would tell some dependency can't be found in jest/build. Running `run `lerna boostrap --hoist` might alter the package-lock.json. 

Please advise how to use lerna to collaborate within the team to make version consistent in the package-lock.json.

**Comment by YousefED** [comment]
I'm running into the same problem as @thinkkevin. Open source dependencies are automatically upgraded, instead of respecting the exact versions locked in `package.lock`. Any guidance on what's the recommended way to prevent this and make sure all team members are using the exact same package versions?

**Comment by WickyNilliams** [comment]
i keep finding that when `lerna bootstrap` is run, the `package-lock` files end up with tabs instead of spaces for indentation. not sure why this is, since npm does not do it itself. npm supposedly detects indentation and follows suit - is lerna doing something that might confuse npm's detection?

**Comment by marr** [comment]
> i keep finding that when `lerna bootstrap` is run, the `package-lock` files end up with tabs instead of spaces for indentation. not sure why this is, since npm does not do it itself. npm supposedly detects indentation and follows suit - is lerna doing something that might confuse npm's detection?

That was fixed in 5.1.5 https://github.com/lerna/lerna/releases/tag/v5.1.5.

Separately, all the package-lock rewriting stuff is a bit confusing. Suggesting running `lerna bootstrap` in lieu of `npm install` is not going to work, because `npm install` is what brings in the `lerna` executable in the first place. A `postinstall` script of `npx lerna bootstrap --ci` would basically be two installs.

Is there a sensible behavior we can expect to avoid all this lockfile madness?

**Comment by JamesHenry** [comment]
Hi Folks 👋

Please take a look at our published roadmap for Lerna v7 here: https://github.com/lerna/lerna/discussions/3410

One of the key items covered at length on there (please do read it for full context) is that now that we find ourselves in late 2022, it no longer makes sense for lerna to supplement package management concerns (such as installation, boostrapping, linking etc) which are covered reliably for monorepo workspaces by the three main package managers: npm, yarn and pnpm. `lerna bootstrap` et al were developed in a completely different era of the JavaScript ecosystem.

If you have any specific concerns please do join in on that discussion, and provide as much context as possible.

Many thanks 🙏
